server/src/main/java/org/opensearch/index/cache/request/ShardRequestCache.java

```diff
@@ -75,7 +78,16 @@ public void onRemoval(long keyRamBytesUsed, BytesReference value, boolean evicte
         if (value != null) {
             dec += value.ramBytesUsed();
         }
-        totalMetric.dec(dec);
+        if ((totalMetric.count() - dec) < 0) {
+            logger.warn(
+                "Ignoring the operation to deduct memory: {} from RequestStats memory_size metric as it will "
+                    + "go negative. Current memory: {}. This is a bug.",
+                dec,
+                totalMetric.count()
+            );
+        } else {
+            totalMetric.dec(dec);
+        }
```

Comment:
In theory, you could still get a race condition that goes negative, but it's much less likely with this change.

If two threads are trying to decrement at the same time, they could both succeed on the `if` check (since neither change is enough to go negative), then both would get applied.

Reply:
Addressed comments.